### PR TITLE
Parse PostgreSQL unique constraint violation messages

### PR DESCRIPTION
**Pull request by chrisandreae**
Make a best effort to extract the constraint name and the violated column and include in metadata

### CODE REVIEW

File: lib/view_model/deserialization_error.rb [hunk 1]
```diff
@@ -306,6 +308,99 @@ def initialize(detail, nodes = [])
...
+          # value list.
+          return nil unless stream.delete_prefix!(DETAIL_INFIX)
+
+          [identifiers, stream]
+        end
+
+        private
+
+        QUOTED_IDENTIFIER   = /\A"(?:[^"]|"")+"/
+        UNQUOTED_IDENTIFIER = /\A(?:\p{Alpha}|_)(?:\p{Alnum}|_)*/
```
Comment (thefloweringash) [inline lib/view_model/deserialization_error.rb]: `/\A(?:\p{Alpha}|_)(?:\p{Alnum}|_)*/` looks like it simplifies to `/\A(?:\p{Alpha}|_)+/`

File: lib/view_model/deserialization_error.rb [hunk 2]
```diff
@@ -306,6 +308,99 @@ def initialize(detail, nodes = [])
...
+        end
+
+        self.new(err.message, constraint, columns, values, nodes)
+      end
+
+      class << self
+        DETAIL_PREFIX = 'Key ('
+        DETAIL_SUFFIX = ') already exists.'
+        DETAIL_INFIX  = ')=('
+        def parse_message_detail(detail)
```
Comment (thefloweringash) [inline lib/view_model/deserialization_error.rb]: This reads fine, seems sensible, but contains a decent amount of logic not covered by tests.

File: lib/view_model/deserialization_error.rb [hunk 3]
```diff
@@ -306,6 +308,99 @@ def initialize(detail, nodes = [])
...
+          stream = detail.dup
+
+          return nil unless stream.delete_prefix!(DETAIL_PREFIX)
+          return nil unless stream.delete_suffix!(DETAIL_SUFFIX)
+
+          # The message should start with an identifier list: pop off identifier
+          # tokens while we can.
+          identifiers = []
+
+          identifier = parse_identifier(stream)
```
Comment (thefloweringash) [inline lib/view_model/deserialization_error.rb]: Awkward to have to unroll the loop body once due to the ordering of conditions. Would you prefer `loop/break`?

File: lib/view_model/deserialization_error.rb (hunk 1 above)
Comment (chrisandreae) [reply lib/view_model/deserialization_error.rb]: Alnum vs Alpha

**Pull request merged**